### PR TITLE
Allow scatter and gather to broadcast

### LINKED ISSUE #9407: [feature request] Make `torch.gather` broadcastable.
Currently, `torch.gather` does not broadcast. For example:
```python
t = torch.tensor([[1,2],[1,2]])
torch.gather(t, 1, torch.tensor([[0,1,0],[0,1,0]]))
```
gives
```
tensor([[ 1,  2,  1],
        [ 1,  2,  1]])
```

But

```python
t = torch.tensor([[1,2],[1,2]])
torch.gather(t, 1, torch.tensor([[0,1,0]]))
```
and
```python
t = torch.tensor([[1,2]])
torch.gather(t, 1, torch.tensor([[0,1,0],[0,1,0]]))
```
gives runtime errors.

It would be more convenient to allow broadcasting on every dimension except the dim specified as the second argument in gather, so that the above two codes can give the same result as the first code.

Does this proposal sound good? If yes, I can work on it.

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Closes: https://github.com/pytorch/pytorch/issues/9407

- [x] gather broadcast
- [x] unit test for gather with broadcast
- [x] scatter broadcast
- [x] unit test for scatter with broadcast
- [x] gather gradients
- [x] unit tests for gradients of gather with broadcast
- [x] scatter gradients
- [x] unit tests for gradients of scatter with broadcast
- [x] explain how broadcasting of scatter and gather works in https://pytorch.org/docs/master/notes/broadcasting.html
- [x] change the docstring of scatter and gather to explain the broadcast

cc: @gchanan

### CODE REVIEW

File: aten/src/ATen/native/Gather.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#include "ATen/ExpandUtils.h"
+
+#include <tuple>
+
+namespace at {
+namespace native{
+
+namespace {
+
+std::tuple<std::vector<int64_t>, std::vector<int64_t> > gather_infer_size(IntList tensor, IntList indices, long dim) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Gather.cpp]: NB: these functions are copied and modified from `ExpandUtils.h`

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -1919,3 +1919,11 @@
 
 - func: meshgrid(TensorList tensors) -> TensorList
   variants: function
+
+- func: gather(Tensor self, int64_t dim, IndexTensor index) -> Tensor
+  variants:
```
Comment (vishwakftw) [inline aten/src/ATen/native/native_functions.yaml]: `variants` are `function` and `method` by default.

Comment (zasdfgbnm) [reply aten/src/ATen/native/native_functions.yaml]: Just fixed, thanks!

**Comment by zasdfgbnm** [comment]
No longer WIP, should be working now.

**Comment by LeanderK** [comment]
Why is this not merged yet? This would really help me!

**Comment by LeanderK** [comment]
I think this pull request may have been forgotten, but It would really, really help me! Right now, I have to resize my tensors every time I use gather, which is really hurting my memory consumption, since i also have gradients flowing through it.

@zasdfgbnm  any chance you could revive the pull-request? I am unfortunately not knowledgable about cpp.

**Comment by zasdfgbnm** [comment]
Hi @LeanderK I would have time to work on this at the end of this month. But I am not sure if there will be someone available to review this. Let me ask.

**Comment by ezyang** [comment]
I can review (or find someone to review).  Sorry about this falling off
the radar initially.

Excerpts from Gao, Xiang's message of 2019-07-09 08:28:06 -0700:
> Hi @LeanderK I would have time to work on this at the end of this month. But I am not sure if there will be someone available to review this. Let me ask.
>

**Comment by vishwakftw** [comment]
#23479 seems to be a PR with the same title.

**Comment by zasdfgbnm** [comment]
@vishwakftw Yes, thanks for the cross-referencing!

**Comment by zasdfgbnm** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/pull/23479
